src/cmds/restic/cmd_forget.go

```diff
@@ -157,51 +118,69 @@ func runForget(opts ForgetOptions, gopts GlobalOptions, args []string) error {
 		Tags:    opts.KeepTags,
 	}
 
-	if policy.Empty() {
-		return nil
-	}
-
-	// then, load all remaining snapshots
 	snapshots, err := restic.LoadAllSnapshots(repo)
 	if err != nil {
 		return err
 	}
 
-	// group by hostname and dirs
+	// Group snapshots by hostname and dirs.
 	type key struct {
 		Hostname string
-		Dirs     string
+		Paths    string
+		Tags     string
 	}
 
-	snapshotGroups := make(map[key]restic.Snapshots)
+	snapshotGroups := make(map[key][]*restic.Snapshot)
 
 	for _, sn := range snapshots {
-		if opts.Hostname != "" && sn.Hostname != opts.Hostname {
+		if opts.Host != "" && sn.Hostname != opts.Host {
 			continue
 		}
 
 		if !sn.HasTags(opts.Tags) {
 			continue
 		}
 
-		k := key{Hostname: sn.Hostname, Dirs: strings.Join(sn.Paths, ":")}
+		if !sn.HasPaths(opts.Paths) {
+			continue
+		}
+
+		tags := ""
+		if opts.GroupByTags {
+			tags = strings.Join(sn.Tags, ":")
+		}
+		k := key{Hostname: sn.Hostname, Tags: tags, Paths: strings.Join(sn.Paths, ":")}
```

Comment:
Re 1: Nothing. It is simply a separator for when the key is printed later on. It will look ugly, sure ^^
Re 2: Sorting, ok. Good catch there. I think the Paths need a sort for the same reason, on the same host I can backup /etc,/home and later on /home,/etc. Unless they get sorted elsewhere.

The JSON representation just for tags is overkill IMO. Either struct equivalence works or it doesn't. If it doesn't we need to do something for the whole struct, not just the tags.